libmscore/treewrite.cpp

```diff
@@ -0,0 +1,262 @@
+
+static bool anyElementsInTrack(Measure* m, int track)
+{
+    for (const Segment& s : m->segments()) {
+        Element* e = s.element(track);
+        if (e) {
+            return true;
+        }
+    }
+    return false;
+}
+
+//---------------------------------------------------------
+//   ScoreElement
+//---------------------------------------------------------
+
+void ScoreElement::treeWrite(XmlWriter& xml)
+{
+    xml.stag(this);
+    for (ScoreElement* ch : *this) {
+        ch->treeWrite(xml);
+    }
+    xml.etag();
+}
+
+//---------------------------------------------------------
+//   Element
+//---------------------------------------------------------
+
+void Element::treeWrite(XmlWriter& xml)
+{
+    if (isTuplet()) {
+        qDebug();
+    }
+    if (generated()) {
+        return;
+    }
+    if (!isUserModified() && !shouldWrite(this)) {
+        return;
+    }
+    xml.stag(this);
+    writeAllProperties(xml, this);
+    for (ScoreElement* ch : *this) {
+        ch->treeWrite(xml);
+    }
+    xml.etag();
+}
+
+//---------------------------------------------------------
+//   for Spanners
+//---------------------------------------------------------
+
+void Spanner::treeWrite(XmlWriter& xml)
+{
+    this->writeSpannerStart(xml, toElement(treeParent()), track());
+}
+
+//---------------------------------------------------------
+//   Score
+//---------------------------------------------------------
+
+void Score::treeWrite(XmlWriter& xml)
+{
+    xml.header();
+    xml.stag("museScore version=\"3.01\"");
+    xml.stag(this);
+    // write all measures in staff 1 first, then staff 2, etc.
+    for (int staffIdx = 0; staffIdx < nstaves(); ++staffIdx) {
+        xml.stag(staff(staffIdx), QString("id=\"%1\"").arg(staffIdx + 1));
+        for (MeasureBase* m = measures()->first(); m != nullptr; m = m->next()) {
+            if (m->isMeasure()) {
+                toMeasure(m)->treeWriteStaff(xml, staffIdx);
+            } else {
+                // non-measure things (like boxes) to be written only once in staff 1
+                if (staffIdx == 0) {
+                    m->treeWrite(xml);
+                }
+            }
+        }
+        xml.etag();
+    }
+    xml.etag(); // score
+    xml.etag(); // musescore
+}
+
+//---------------------------------------------------------
+//   Measure::treeWriteStaff
+///   Writes one staff in one measure
+//---------------------------------------------------------
+
+void Measure::treeWriteStaff(XmlWriter& xml, int staffIdx)
+{
+    xml.stag(this);
+    for (Element* e : el()) {
+        e->treeWrite(xml);
+    }
+    // write voice 1 first, then voice 2, .. upto VOICES
+    for (int voice = 0; voice < VOICES; voice++) {
+        int track = staffIdx * VOICES + voice;
+        xml.setCurTrack(track);
+        // check if voice empty?
+        if (!anyElementsInTrack(this, track)) {
+            continue;
+        }
+        xml.stag("voice");
```

Comment:
What happens if Voice 3 has notes but Voice 2 was empty? The current format inserts empty`<voice/>` tags for empty voices up to the last non-empty voice.

You could keep an `insertEmptyVoices` counter and insert that many voices before starting a non-empty voice.

```c++
for (; insertEmptyVoices > 0; --insertEmptyVoices) {
    xml.tagE("voice");
}
```

There should probably be a test for this somewhere. There should be one measure with voice 2 empty, one with voice 3 empty, and one with voice 2 and 3 empty. I don't think voice 1 can be empty, but if it can then that should be tested too.

Reply:
Oh right, I didn't realise that the current format also has tags for empty voices. I'll correct this.